dataflow/spanner-io/src/test/java/com/example/dataflow/SpannerGroupWriteIT.java

```diff
@@ -0,0 +1,149 @@
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.annotation.Nullable;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+@SuppressWarnings("checkstyle:abbreviationaswordinname")
+public class SpannerGroupWriteIT {
+
+  String instanceId;
+  String databaseId;
+
+  Path tempPath;
+  Spanner spanner;
+  SpannerOptions spannerOptions;
+
+  @Before
+  public void setUp() throws Exception {
+    instanceId = System.getProperty("spanner.test.instance");
+    databaseId = "df-spanner-groupwrite-it";
+
+    spannerOptions = SpannerOptions.getDefaultInstance();
+    spanner = spannerOptions.getService();
+
+    DatabaseAdminClient adminClient = spanner.getDatabaseAdminClient();
+
+    try {
+      adminClient.dropDatabase(instanceId, databaseId);
+    } catch (SpannerException e) {
+      // Does not exist, ignore.
+    }
+
+    Operation<Database, CreateDatabaseMetadata> op = adminClient
+        .createDatabase(instanceId, databaseId, Arrays.asList("CREATE TABLE users ("
+                + "id STRING(MAX) NOT NULL, state STRING(MAX) NOT NULL) PRIMARY KEY (id)",
+            "CREATE TABLE PendingReviews (id INT64, action STRING(MAX), "
+                + "note STRING(MAX), userId STRING(MAX),) PRIMARY KEY (id)"));
+
+    op.waitFor();
+
+    DatabaseClient dbClient = getDbClient();
+
+    List<Mutation> mutations = new ArrayList<>();
+    for (int i = 0; i < 20; i++) {
+      mutations.add(
+          Mutation.newInsertBuilder("users").set("id").to(Integer.toString(i)).set("state")
+              .to("ACTIVE").build());
+    }
+    TransactionRunner runner = dbClient.readWriteTransaction();
+    runner.run(new TransactionRunner.TransactionCallable<Void>() {
+
+      @Nullable
+      @Override
+      public Void run(TransactionContext tx) throws Exception {
+        tx.buffer(mutations);
+        return null;
+      }
+    });
+
+    String content = IntStream.range(0, 10).mapToObj(Integer::toString)
+        .collect(Collectors.joining("\n"));
+    tempPath = Files.createTempFile("suspicious-ids", "txt");
+    Files.write(tempPath, content.getBytes());
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    DatabaseAdminClient adminClient = spanner.getDatabaseAdminClient();
+    try {
+      adminClient.dropDatabase(instanceId, databaseId);
+    } catch (SpannerException e) {
+      // Failed to cleanup.
+    }
+
+    spanner.close();
+  }
+
+  @Test
+  public void testEndToEnd() throws Exception {
+    SpannerGroupWrite.main(
+        new String[] { "--instanceId=" + instanceId, "--databaseId=" + databaseId,
+            "--suspiciousUsersFile=" + tempPath, "--runner=DirectRunner" });
+
+    DatabaseClient dbClient = getDbClient();
+    try (ReadContext context = dbClient.singleUse()) {
+      ResultSet rs = context.executeQuery(
+          Statement.newBuilder("SELECT COUNT(*) FROM users WHERE STATE = @state").bind("state")
```

Comment:
?? Why do you bother to bind to a constant? ??

(I'm ok w/ it as long as you had a reason)

Reply:
It's just a good practice, if we want to change this to a variable in future